go/vt/srvtopo/server.go

```diff
@@ -47,3 +47,71 @@ type Server interface {
 	// a new value or an error occurs.
 	WatchSrvVSchema(ctx context.Context, cell string, callback func(*vschemapb.SrvVSchema, error))
 }
+
+// NewKeyspaceFilteringServer constructs a new server based on the provided
+// implementation that prevents the specified keyspaces from being exposed
+// to consumers of the new Server.
+func NewKeyspaceFilteringServer(underlying Server, selectedKeyspaces []string) Server {
+	keyspaces := map[string]bool{}
+	for _, ks := range selectedKeyspaces {
+		keyspaces[ks] = true
+	}
+
+	return keyspaceFilteringServer{
+		server:         underlying,
+		selectKeyspace: keyspaces,
+	}
+}
+
+type keyspaceFilteringServer struct {
+	server         Server
+	selectKeyspace map[string]bool
+}
+
+func (ksf keyspaceFilteringServer) GetTopoServer() *topo.Server {
+	return ksf.server.GetTopoServer()
+}
+
+func (ksf keyspaceFilteringServer) GetSrvKeyspaceNames(
+	ctx context.Context,
+	cell string,
+) ([]string, error) {
+	keyspaces, err := ksf.server.GetSrvKeyspaceNames(ctx, cell)
+	ret := make([]string, 0, len(keyspaces))
+	for _, ks := range keyspaces {
+		if ksf.selectKeyspace[ks] {
+			ret = append(ret, ks)
+		}
+	}
+	return ret, err
+}
+
+func (ksf keyspaceFilteringServer) GetSrvKeyspace(
+	ctx context.Context,
+	cell,
+	keyspace string,
+) (*topodatapb.SrvKeyspace, error) {
+	if !ksf.selectKeyspace[keyspace] {
+		return nil, topo.NewError(topo.NoNode, keyspace)
+	}
+
+	return ksf.server.GetSrvKeyspace(ctx, cell, keyspace)
+}
+
+func (ksf keyspaceFilteringServer) WatchSrvVSchema(
+	ctx context.Context,
+	cell string,
+	callback func(*vschemapb.SrvVSchema, error),
+) {
+	filteringCallback := func(schema *vschemapb.SrvVSchema, err error) {
+		for ks := range schema.Keyspaces {
```

Comment:
For safety you need a check that `schema` isn't nil here.

Reply:
✅; thank you, good catch